spec/abstract/SchedContext_A.thy

```diff
@@ -195,72 +190,78 @@ where
+       then let new_hd = \<lparr>r_time = r_time r1 - r_amount r0,
+                          r_amount = r_amount r0 + r_amount r1\<rparr>
+            in MIN_BUDGET_merge (new_hd # rs)
+       else r0 # r1 # rs)"
+by pat_completeness auto
+
 definition
   refill_budget_check :: "ticks \<Rightarrow> (unit, 'z::state_ext) s_monad"
 where
   "refill_budget_check usage = do
     sc_ptr \<leftarrow> gets cur_sc;
     sc \<leftarrow> get_sched_context sc_ptr;
     ready \<leftarrow> get_sc_refill_ready sc_ptr;
-    period \<leftarrow> return $ sc_period sc;
+    period \<leftarrow> return (sc_period sc);
+    refills \<leftarrow> return (sc_refills sc);
+
     robin \<leftarrow> is_round_robin sc_ptr;
     assert (\<not>robin);
-    refills \<leftarrow> return (sc_refills sc);
 
-    last_entry \<leftarrow> return $ r_time (hd refills);
-
-    used \<leftarrow> return $ \<lparr>r_time = last_entry + period, r_amount = usage\<rparr>;
-
-    if \<not>ready \<or> r_amount (hd refills) < usage
-    then set_refills sc_ptr [\<lparr>r_time = last_entry + period + usage, r_amount = sc_budget sc\<rparr>]
-    else if usage = r_amount (hd refills)
-         then set_refills sc_ptr (schedule_used False (tl refills) used)
-                \<comment> \<open>if refills has length at most @{text \<open>sc_refills_max sc\<close>}, then popping the head
-                    will ensure the refills are not full, so we may use False here\<close>
-         else do remnant \<leftarrow> return $ r_amount (hd refills) - usage;
-                 if remnant < MIN_BUDGET
-                 then if tl refills = []
-                      then set_refills sc_ptr [\<lparr>r_time = last_entry + period - remnant,
-                                                r_amount = r_amount (hd refills)\<rparr>]
-                      else do
-                        new_snd \<leftarrow> return $ (\<lparr>r_time = r_time (hd (tl refills)) - remnant,
-                                              r_amount = r_amount (hd (tl refills)) + remnant \<rparr>);
-                        set_refills sc_ptr (schedule_used False (new_snd # tl (tl refills)) used)
-                      od
-                 else do
-                   full \<leftarrow> refill_full sc_ptr;
-                   rfhd \<leftarrow> return $ (hd refills);
-                   new_head \<leftarrow> return $ (\<lparr>r_time = r_time rfhd + usage, r_amount = remnant\<rparr>);
-                   set_refills sc_ptr (schedule_used full (new_head # (tl refills)) used)
-                 od
-              od
+    usage' \<leftarrow> return $ min usage (r_amount (hd refills));
+
+    if usage' > 0
+    then do used \<leftarrow> return \<lparr>r_time = r_time (hd refills) + period, r_amount = usage'\<rparr>;
+            full \<leftarrow> refill_full sc_ptr;
+            refills' \<leftarrow> return $ schedule_used full refills used;
+            usage'' \<leftarrow> return $ min usage' (r_amount (hd refills'));
```

Comment:
Why does this need to be recalculated here?

Reply:
Actually, I think this line should be removed and I should be using `usage'` here. Thanks!

Reply:
Wait, I was just using `usage''` as a way to calculate the `min` that's in the C after the `schedule_used`. I could lump this all into the following line if it makes it easier to read or understand

Reply:
I think I'd at least rename this from `usage''` to `min`, or something, to be a little bit less confusing

Reply:
That looks like an error on my part, only one MIN should be there

Reply:
The second one should be redundant. I'll remove it in the C.